src/main/kotlin/org/rust/lang/core/resolve/indexes/RsImplIndex.kt

```diff
@@ -47,8 +43,12 @@ class RsImplIndex : AbstractStubIndex<RustTypeFingerprint, RsImplItem>() {
                 RsImplItem::class.java
             ).asSequence()
             return when {
-                else -> elements.filter {
-                    canUnify(target, it.typeReference?.type ?: RustUnknownType)
+                else -> elements.filter { impl ->
+                    val ty = impl.typeReference?.type
+                    // Addition class check is a temporal solution to filter impls for type parameter
+                    // with the same name
+                    // struct S; impl<S: Tr1> Tr2 for S {}
+                    ty != null && ty.javaClass == target.javaClass && ty.isAssignableFrom(target, project)
```

Comment:
That's unlikely to bite us in practice, so I suggest to remove this special case and instead write a test that verifies current  wrong behavior with a `FIXME` mark. 

Ideally, purely generic impls of this form should be handled separately. We should collect them in a separate index (maybe keyed by the trait) and then check if they are implemented for particular type.